File: platform/ios/xcode/FullStacked/InstanceEditor.swift
//
//  InstanceEditor.swift
//  FullStacked
//
//  Created by Charles-Philippe Lepage on 2024-03-22.
//

import Foundation
import SwiftyJSON
import UIKit

let paths = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true);
let documentsDirectory = paths.first!

class InstanceEditor: Instance {
    static var singleton: InstanceEditor?
    
    init(){
        let editorDirectory = Bundle.main.path(forResource: "build", ofType: nil)!
        super.init(adapter: AdapterEditor(baseDirectory: editorDirectory))
        self.webview.isOpaque = false
        InstanceEditor.singleton = self
    }
}

class AdapterEditor: Adapter {
    let rootDirectory = documentsDirectory
    private let baseJSFile = Bundle.main.path(forResource: "base", ofType: "js", inDirectory: "js")!
    let cacheDirectory = FileManager.default.temporaryDirectory.absoluteString
    let configDirectory = ".config/fullstacked"
    let nodeModulesDirectory: String
    let fsEditor: AdapterFS
    let bonjour = Bonjour()
    let multipeer = Multipeer()
    
    init(baseDirectory: String) {
        self.nodeModulesDirectory = configDirectory + "/node_modules"
        self.fsEditor = AdapterFS(baseDirectory: self.rootDirectory);
        super.init(projectId: nil, baseDirectory: baseDirectory)
        
        
        self.bonjour.onPeerNearby = {eventType, peerNearbyBonjour in
            let message = [
                "eventType": eventType,
                "peerNearby": [
                    "type": 1,
                    "peer": [
                        "id": peerNearbyBonjour.id,
                        "name": peerNearbyBonjour.name
                    ],
                    "addresses": peerNearbyBonjour.addresses,
                    "port": peerNearbyBonjour.port
                ]
            ]
            InstanceEditor.singleton?.push(messageType: "peerNearby", message: JSON(message).rawString()!)
        }
        self.multipeer.onPeerNearby = {eventType, peerNearbyMultipeer in
            let message = [
                "eventType": eventType,
                "peerNearby": [
                    "type": 2,
                    "peer": [
                        "id": peerNearbyMultipeer.peer.id,
                        "name": peerNearbyMultipeer.peer.name
                    ],
                    "id": peerNearbyMultipeer.id
                ]
            ]
            InstanceEditor.singleton?.push(messageType: "peerNearby", message: JSON(message).rawString()!)
        }
        
        self.multipeer.onPeerConnection = { id, type, state in
            let message = [
                "id": id,
                "type": type,
                "state": state
            ]
            InstanceEditor.singleton?.push(messageType: "peerConnection", message: JSON(message).rawString()!)
        }
        
        self.multipeer.onPeerData = { id, data in
            let message = [
                "id": id,
                "data": data
            ]
            InstanceEditor.singleton?.push(messageType: "peerData", message: JSON(message).rawString()!)
        }
    }
    
    override func callAdapterMethod(methodPath: [String.SubSequence], body: Data, done: @escaping (_ maybeData: Any?) -> Void) {
        if(methodPath.count == 0) {
            return done(nil)
        }
        
        let json = try! JSON(data: body)
        
        let writeFile = { (path: String, data: Data, recursive: Bool) in
            if(recursive) {
                let directory = path.split(separator: "/").dropLast()
                self.fsEditor.mkdir(path: directory.joined(separator: "/"))
            }
            
            return self.fsEditor.writeFile(file: path, data: data)
        }
        
        switch(methodPath.first) {
            case "directories":
                switch(methodPath[1]) {
                    case "rootDirectory": return done(self.rootDirectory)
                    case "cacheDirectory": return done(self.cacheDirectory)
                    case "configDirectory": return done(self.configDirectory)
                    case "nodeModulesDirectory": return done(self.nodeModulesDirectory)
                    default: break
                }
                break
            case "fs":
                if (json[1]["absolutePath"].boolValue || json[2]["absolutePath"].boolValue) {
                    switch(methodPath[1]){
                        case "readFile": return done(self.fsEditor.readFile(path: json[0].stringValue, utf8: json[1]["encoding"].stringValue == "utf8"))
                        case "writeFile":
                            var data: Data;
                            
                            if(json[1]["type"].stringValue == "Uint8Array") {
                                let uint8array = json[1]["data"].arrayValue.map({ number in
                                    return number.uInt8!
                                })
                                data = Data(uint8array)
                            } else {
                                data = json[1].stringValue.data(using: .utf8)!
                            }
                        
                            return done(writeFile(json[0].stringValue, data, json[2]["recursive"].boolValue))
                        case "writeFileMulti":
                            for fileJSON in json[0].arrayValue {
                                var data: Data;
                                
                                if(fileJSON["data"]["type"].stringValue == "Uint8Array") {
                                    let uint8array = fileJSON["data"]["data"].arrayValue.map({ number in
                                        return number.uInt8!
                                    })
                                    data = Data(uint8array)
                                } else {
                                    data = fileJSON["data"].stringValue.data(using: .utf8)!
                                }
                                
                                let maybeError = writeFile(fileJSON["path"].stringValue, data, json[1]["recursive"].boolValue)
                                if(maybeError is AdapterError){
                                    return done(maybeError)
                                }
                            }
                            return done(true)
                        case "unlink": return done(self.fsEditor.unlink(path: json[0].stringValue))
                        case "readdir": return done(self.fsEditor.readdir(path: json[0].stringValue, withFileTypes: json[1]["withFileTypes"].boolValue, recursive: json[1]["recursive"].boolValue))
                        case "mkdir": return done(self.fsEditor.mkdir(path: json[0].stringValue))
                        case "rmdir": return done(self.fsEditor.rmdir(path: json[0].stringValue))
                        case "stat": return done(self.fsEditor.stat(path: json[0].stringValue))
                        case "lstat": return done(self.fsEditor.lstat(path: json[0].stringValue))
                        case "exists":
                            let exists = self.fsEditor.exists(path: json[0].stringValue)
                            return done(exists == nil ? false : exists)
                        default: break;
                    }
                }
                break
            case "esbuild":
                switch(methodPath[1]) {
                case "baseJS":
                    let content = FileManager.default.contents(atPath: self.baseJSFile)!
                    return done(String(data: content, encoding: .utf8)!)
                case "check": return done("1")
                case "install": break
                case "tmpFile":
                    switch(methodPath[2]) {
                    case "write":
                        let path = self.cacheDirectory + json[0].stringValue
                        let data = json[1].stringValue.data(using: .utf8)!
                        try! data.write(to: URL(string: path)!)
                        return done(String(path.dropFirst("file://".count)))
                    case "unlink":
                        let path = self.cacheDirectory + json[0].stringValue
                        try! FileManager.default.removeItem(at: URL(string: path)!)
                        return done(true)
                    default: break;
                    }
                case "build":
                    let inputPtr = UnsafeMutablePointer<Int8>(mutating: (json[0].stringValue as NSString).utf8String)
                    let outPtr = UnsafeMutablePointer<Int8>(mutating: ("index" as NSString).utf8String)
                    let outdirPtr = UnsafeMutablePointer<Int8>(mutating: (json[1].stringValue as NSString).utf8String)
                    let nodePathPtr = UnsafeMutablePointer<Int8>(mutating: (self.rootDirectory + "/" + self.nodeModulesDirectory as NSString).utf8String)
                    
                    var errorsPtr = UnsafeMutablePointer<Int8>(nil)
                    
                    build(inputPtr,
                          outPtr,
                          outdirPtr,
                          nodePathPtr,
                          &errorsPtr)
                                
                    let errors = String.init(cString: errorsPtr!, encoding: .utf8)!
                    if(!errors.isEmpty) {
                        return done(JSON(parseJSON: errors))
                    }
                
                    return done(true)
                default: break
                }
                break
            case "run":
                let projectDirectory = self.rootDirectory + "/" + json[0]["location"].stringValue
                    
                let runningInstance = RunningInstances.singleton?.getInstance(projectDirectory: projectDirectory)
            
                if(runningInstance != nil) {
                    runningInstance!.webview.reload()
                } else {
                    let project = Project(
                        location: projectDirectory,
                        id: json[0]["id"].stringValue,
                        title: json[0]["title"].stringValue)
                    RunningInstances.singleton?.addInstance(instance: Instance(project: project))
                }
            
                return done(true)
            case "open": 
                let projectLocation = self.rootDirectory + "/" + json[0]["location"].stringValue
                UIApplication.shared.open(URL(string: "shareddocuments://" + projectLocation)!)
                return done(true)
            case "connectivity":
                switch(methodPath[1]) {
                case "infos":
                    return done(false)
                case "name":
                    return done(UIDevice.current.name)
                case "peers":
                    switch(methodPath[2]){
                    case "nearby":
                        var peersNearby: [JSON] = []
                        let peersNearbyBonjour = self.bonjour.getPeersNearby().arrayValue
                        let peersNearbyMultipeer = self.multipeer.getPeersNearby().arrayValue
                        
                        for peerNearby in peersNearbyBonjour {
                            peersNearby.append(peerNearby)
                        }
                        for peerNearby in peersNearbyMultipeer {
                            peersNearby.append(peerNearby)
                        }
                        
                        return done(JSON(peersNearby))
                    default: break
                    }
                case "advertise": 
                    switch(methodPath[2]){
                    case "start": 
                        self.multipeer.startAdvertising(id: json[0]["id"].stringValue, name: json[0]["name"].stringValue)
                        return done(true)
                    case "stop": 
                        self.multipeer.stopAdvertising()
                        return done(true)
                    default: break
                    }
                case "browse": 
                    switch(methodPath[2]){
                    case "start":
                        self.bonjour.startBrowsing()
                        self.multipeer.startBrowsing()
                        return done(true)
                    case "peerNearbyIsDead":
                        self.bonjour.peerNearbyIsDead(id: json[0].stringValue)
                        return done(true)
                    case "stop":
                        self.bonjour.stopBrowsing()
                        self.multipeer.stopBrowsing()
                        return done(true)
                    default: break
                    }
                case "open":
                    self.multipeer.open(id: json[0].stringValue, meId: json[1]["id"].stringValue, meName: json[1]["name"].stringValue)
                    return done(true)
                case "disconnect":
                    self.multipeer.disconnect(id: json[0].stringValue)
                    return done(true)
                case "trustConnection":
                    self.multipeer.trustConnection(id: json[0].stringValue)
                    return done(true)
                case "send":
                    self.multipeer.send(id: json[0].stringValue, data: json[1].stringValue, pairing: json[2].boolValue)
                    return done(true)
                case "convey":
                    let data = json[1].stringValue;
                    let projectId = json[0].stringValue;
                    RunningInstances.singleton!.instances.forEach({instance in
                        if(instance.adapter.projectId == projectId) {
                            instance.push(messageType: "peerData", message: data)
                        }
                    })
                    return done(true)
                default: break
                }
            default: break
        }
        
        return super.callAdapterMethod(methodPath: methodPath, body: body, done: done)
    }
}

File: platform/android/studio/app/src/main/java/org/fullstacked/editor/InstanceEditor.kt
package org.fullstacked.editor

import android.content.Intent
import android.os.Environment
import android.provider.DocumentsContract
import org.fullstacked.editor.connectivity.Bonjour
import org.fullstacked.editor.connectivity.Peer
import org.fullstacked.editor.connectivity.PeerNearby
import org.fullstacked.editor.connectivity.WSS
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.io.InputStream


val editorProject = Project(
    location = "",
    id = "org.fullstacked.editor",
    title = "FullStacked"
)

class InstanceEditor(var context: MainActivity) : Instance(
    project = editorProject,
    init = false
) {
    companion object {
        lateinit var singleton: InstanceEditor
        fun initialized(): Boolean = this::singleton.isInitialized
    }

    val instances = mutableListOf<Instance>()

    init {
        singleton = this

        this.adapter = AdapterEditor(
            projectId = editorProject.id,
            baseDirectory = this.context.filesDir.toString()
        )

        this.render()
    }

    override fun render() {
        val webView = createWebView(
            ctx = this.context,
            adapter = this.adapter,
            isEditor = true
        )

        if(this.webViewState != null) {
            webView.restoreState(this.webViewState!!)
        }

        this.webViewId = webView.id

        this.context.setContentView(webView)
    }
}

class AdapterEditor(
    projectId: String,
    private val baseDirectory: String,
): Adapter(projectId, baseDirectory) {
    private val webSocketServer = WSS()
    private val bonjour = Bonjour(webSocketServer)

    companion object {
        init {
            System.loadLibrary("editor")
        }
    }

    private external fun build(
        input: String,
        outdir: String,
        nodePath: String,
    ): String

    override fun getFile(path: String): InputStream? {
        return try {
            InstanceEditor.singleton.context.assets.open(path)
        } catch (e: IOException) {
            null
        }
    }

    private fun directoriesSwitch(directory: String): String? {
        when (directory) {
            "rootDirectory" -> return this.baseDirectory
            "cacheDirectory" -> return InstanceEditor.singleton.context.cacheDir.toString()
            "configDirectory" -> return ".config"
            "nodeModulesDirectory" -> return ".cache/node_modules"
        }

        return null
    }

    private fun esbuildSwitch(methodPath: List<String>, body: String?) : Any? {
        when (methodPath.first()) {
            "check" -> return true
            "baseJS" -> return convertInputStreamToString(InstanceEditor.singleton.context.assets.open("base.js"))
            "tmpFile" -> {
                when (methodPath.elementAt(1)) {
                    "write" -> {
                        val json = JSONArray(body)
                        val filePath = InstanceEditor.singleton.context.cacheDir.toString() + "/" + json.getString(0)
                        val file = File(filePath)
                        file.writeText(json.getString(1), Charsets.UTF_8)
                        return filePath
                    }
                    "unlink" -> {
                        val json = JSONArray(body)
                        val filePath = InstanceEditor.singleton.context.cacheDir.toString() + "/" + json.getString(0)
                        val file = File(filePath)
                        file.delete()
                        return true
                    }
                }
            }
            "build" -> {
                val json = JSONArray(body)

                val errors = build(
                    input = json.getString(0),
                    outdir = json.getString(1),
                    nodePath = this.baseDirectory + "/.cache/node_modules"
                )

                if (errors.isEmpty())
                    return true

                return JSONArray(errors)
            }
        }

        return null
    }

    private fun connectivitySwitch(methodPath: List<String>, body: String?) : Any? {
        when (methodPath.first()) {
            "name" -> return android.os.Build.MODEL
            "infos" -> {
                val infos = JSONObject()
                infos.put("port", this.webSocketServer.port)

                val addresses = Bonjour.getIpAddress()
                val infoAddresses = JSONArray()
                for (address in addresses)
                    infoAddresses.put(address)

                val networkInterface = JSONObject()
                networkInterface.put("name", "Active Network")
                networkInterface.put("addresses", infoAddresses)

                val networkInterfaces = JSONArray()
                networkInterfaces.put(networkInterface)

                infos.put("networkInterfaces", networkInterfaces)

                return infos
            }
            "peers" -> {
                when (methodPath.elementAt(1)) {
                    "nearby" -> {
                        val json = JSONArray()

                        this.bonjour.getPeersNearby().forEach { peerNearby ->
                            json.put(Bonjour.serializePeerNearby(peerNearby)) }

                        return json
                    }
                }
            }
            "advertise" -> {
                when (methodPath.elementAt(1)) {
                    "start" -> {
                        val json = JSONArray(body)
                        val peerJSON = json.getJSONObject(0)
                        this.bonjour.startAdvertising(Peer(
                            id = peerJSON.getString("id"),
                            name = peerJSON.getString("name")
                        ))
                        return true
                    }
                    "stop" -> {
                        this.bonjour.stopAdvertising()
                        return true
                    }
                }
            }
            "browse" -> {
                when (methodPath.elementAt(1)) {
                    "start" -> {
                        this.bonjour.startBrowsing()
                        return true
                    }
                    "peerNearbyIsDead" -> {
                        val args = JSONArray(body)
                        this.bonjour.peerNearbyIsDead(args.getString(0))
                        return true
                    }
                    "stop" -> {
                        this.bonjour.stopBrowsing()
                        return true
                    }
                }
            }
            "open" -> {
                val peerNearbyJSON = JSONArray(body).getJSONObject(0)
                val peerJSON = peerNearbyJSON.getJSONObject("peer")
                val peerNearby = PeerNearby(
                    peer = Peer(
                        id = peerJSON.getString("id"),
                        name = peerJSON.getString("name")
                    ),
                    addresses = listOf(peerNearbyJSON.getJSONArray("addresses").getString(0)),
                    port = peerNearbyJSON.getString("port").toInt()
                )

                return true
            }
            "disconnect" -> {
                val json = JSONArray(body)
                this.webSocketServer.disconnect(json.getString(0))
                return true
            }
            "trustConnection" -> {
                val json = JSONArray(body)
                this.webSocketServer.trustConnection(json.getString(0))
                return true
            }
            "send" -> {
                val json = JSONArray(body);
                this.webSocketServer.send(json.getString(0), json.getString(1), json.getBoolean(2))
                return true
            }
            "convey" -> {
                var projectId = ""
                var data = ""

                val args = JSONArray(body)
                try{
                    projectId = args.getString(0)
                }catch (_: Exception) { }
                try{
                    data = args.getString(1)
                }catch (_: Exception) { }

                InstanceEditor.singleton.instances.forEach { instance ->
                    if(instance.project.id == projectId) {
                        instance.push("peerData", data)
                    }
                }

                return true
            }
        }
        return null
    }

    private fun run(project: JSONObject): Boolean {
        InstanceEditor.singleton.context.runOnUiThread {
            val instance = Instance(
                project = Project(
                    id = project.getString("id"),
                    title = project.getString("title"),
                    location = project.getString("location")
                )
            )
            InstanceEditor.singleton.instances.add(instance)
        }

        return true
    }

    override fun callAdapterMethod(methodPath: ArrayList<String>, body: String?): Any? {
        when (methodPath.first()) {
            "directories" -> return this.directoriesSwitch(methodPath.elementAt(1))
            "esbuild" -> return this.esbuildSwitch(methodPath.subList(1, methodPath.size), body)
            "connectivity" -> return this.connectivitySwitch(methodPath.subList(1, methodPath.size), body)
            "run" -> return this.run(JSONArray(body).getJSONObject(0))
            "open" -> {
                val projectJSON = JSONArray(body).getJSONObject(0)
                val title = projectJSON.getString("title")
                val location = projectJSON.getString("location")

                val file = File(this.baseDirectory + "/" + location + "/" + title + ".zip")
                if(!file.exists()) return false

                val out = File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).absolutePath + "/" + title + ".zip")
                file.copyTo(out, true)

                val intent = Intent(Intent.ACTION_VIEW)
                intent.setType(DocumentsContract.Document.MIME_TYPE_DIR)
                InstanceEditor.singleton.context.startActivity(intent)
                return true
            }
            "fs" -> {
                var absolutePath = false
                var utf8 = false
                val json = if (!body.isNullOrEmpty()) JSONArray(body) else JSONArray("[]")

                // writeFile
                if (json.length() > 2) {
                    try {
                        val opt = JSONObject(json.getString(2))
                        absolutePath = opt.getBoolean("absolutePath")
                    } catch (_: Exception) {
                    }
                }
                // readFile, writeFileMulti
                else if (json.length() > 1) {
                    try {
                        val opt = JSONObject(json.getString(1))
                        absolutePath = opt.getBoolean("absolutePath")
                    } catch (_: Exception) {
                    }
                    try {
                        val opt = JSONObject(json.getString(1))
                        utf8 = opt.getString("encoding") == "utf8"
                    } catch (_: Exception) {
                    }
                }

                if (absolutePath) return super.callAdapterMethod(methodPath, body)

                if (json.length() == 0) return null

                val file = this.getFile(json.getString(0))

                if (file != null && utf8) {
                    return convertInputStreamToString(file)
                }

                return file
            }
        }


        return super.callAdapterMethod(methodPath, body)
    }
}


fun convertInputStreamToString(inputStream: InputStream): String {
    val result = ByteArrayOutputStream()
    val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
    var length: Int
    while ((inputStream.read(buffer).also { length = it }) != -1) {
        result.write(buffer, 0, length)
    }

    return result.toString("UTF-8")
}